### ISSUE TITLE
WebSockets

### ISSUE THREAD

**Issue by danroth27** [aspnet/Docs#204]
Fundamentals -> WebSockets

**Comment by dlebedynskyi** [comment]
@danroth27, any news on webSockets for core? I see it is in rc1 so at least some basic example or api doc  would be real great.

**Comment by spboyer** [comment]
* https://www.nuget.org/packages/Microsoft.AspNetCore.WebSockets/
*  [samples folder in the WebSockets repository.](https://github.com/aspnet/WebSockets/tree/dev/samples)

```csharp
app.Use(async (context, next) =>
{
  if (context.WebSockets.IsWebSocketRequest)
  {
    var webSocket = await context.WebSockets.AcceptWebSocketAsync();
    await DoSomethingCool(context, webSocket);
  }
  else
  {
    await next();
  }
});
```

@danroth27 : getting requests for priority on this

**Comment by Rick-Anderson** [comment]
see also #2789

**Comment by danroth27** [comment]
We have https://docs.microsoft.com/en-us/aspnet/core/aspnetcore-1.1#websockets-support, but we need to figure out where this should go in the actual TOC.

**Comment by danroth27** [comment]
Let's put it under Fundamentals.

**Comment by tdykstra** [comment]
Proposed outline for the doc:

* What is it:
  * Brief explanation of what WebSocket protocol is, [link out for details](https://en.wikipedia.org/wiki/WebSocket). 
  * WS support for ASP.NET Core is provided in [Microsoft.AspNetCore.WebSockets](https://www.nuget.org/packages/Microsoft.AspNetCore.WebSockets/) NuGet package that includes client and server integration components.
  
* Prerequisites
  * ASP.NET Core 1.1
  * Works on all OSes that ASP.NET Core runs on (Windows 7+, Windows Server 2008+, Linux, macOS)
    * Exception: With Kestrel and IIS, or with WebListener, requires Windows 8+ or Windows Server 2012+
  * Supported browsers:  http://caniuse.com/#feat=websockets

* When to use it
  * If you can run on full .NET, SignalR is a better choice for most apps: higher level of abstraction with features such as:
      * Support for a broader range of browser versions by using automatic fallback to alternative transport methods such as server-sent events and long polling.
      * Automatic reconnection when a connection drops.
      * Support for clients calling methods on server or vice versa
      * Support for scaling to multiple servers
  * SignalR is being ported to Core; see [the repo](https://github.com/aspnet/WebSockets) for more info
  * Use WS if you need to work directly with a socket connection. For example, for games that need best possible performance  

* How to use it
  * Install package
  * app.UseWebSockets in Configure method
    * Can configure some settings 
      * KeepAliveInterval
      * ReceiveBufferSize
  * Somewhere later (in Configure method, in an MVC action, etc.) check if it's a WebSocket request and accept the WebSocket request
    * (snippet from sample app)
    * You get back a WebSocket object - [WebSocket](https://docs.microsoft.com/en-us/dotnet/core/api/system.net.websockets.websocket) is  a .NET Core type, in System.Net.WebSockets
    * The TCP connection is upgraded to a WebSocket connection.
  * Once you have the WebSocket object use it to send and receive messages.
    * (snippet from sample app)
    * While doing that you stay where you're at in the Middleware pipeline
  * When done, close the connection, and the rest of the Middleware pipeline runs
    * (snippet from sample app)

* Sample app - use the [echo example from the repo](https://github.com/aspnet/WebSockets/tree/dev/samples/EchoApp), simplified by removing functionality added there for testing.

Please review - @anurse @Rick-Anderson @dlebedynskyi @danroth27

**Comment by anurse** [comment]
Brief comments:

* `Microsoft.AspNetCore.WebSockets` is just the **server** components. Client components are part of the .NET Standard Library
* SignalR does not work on ASP.NET Core at all, even on full .NET. I'd avoid recommending the old SignalR at all in this doc. What we can do is mention that SignalR provides a richer application model, but is not supported on ASP.NET Core yet, and link off to the repo for more information on the future of SignalR on ASP.NET Core
  * SignalR Repo is actually at https://github.com/aspnet/SignalR
* Short descriptions of the options you can provide in `app.UseWebSockets()`:
  * `KeepAliveInterval` - How frequently we'll send "Ping" frames to the client, to ensure Proxies keep the connection open
  * `ReceiveBufferSize` - The size of the buffer we'll use to receive data. Only a very advanced user would need to change this, for performance tuning based on the size of their data.
* I'd reword the comment about the middleware pipeline something like this: "When you accept the `WebSocket`, the middleware pipeline ends; upon closing the the socket, the pipeline will unwind" (i.e. we stop moving forward in the pipeline when you accept a WebSocket, just as we would when you hit an MVC action, etc., but when you finish, we'll go back up the pipeline)

**Comment by tdykstra** [comment]
@anurse Thanks! Revised the outline.
* Regarding the client components -- the [WebSockets repo readme](https://github.com/aspnet/WebSockets/blob/dev/README.md) says it has "a managed implementation of the WebSocket protocol, along with client and server integration components." What does client components refer to here?
* I'm thinking it's still worth explaining briefly what richer app model means by offering a few examples of SignalR features, but I could be persuaded otherwise.

**Comment by anurse** [comment]
* Good catch, the WebSockets README is out of date, I'll fix it.
* I agree that it's good to explain what SignalR can do, but I don't want to create the wrong impression that the old SignalR can be used with ASP.NET Core, so as long as we're clear that SignalR does not currently work with ASP.NET Core, but we're working on it, I'm good with it
